### PR TITLE
counsel.el (counsel-colors--name-to-hex): protect against invalid input.

### LINKED ISSUE #1853: Error in counsel-color-emacs
GNU Emacs 26.1
macOS Mojave

```elisp
Debugger entered--Lisp error: (wrong-number-of-arguments (3 . 4) 0)
  color-rgb-to-hex()
  apply(color-rgb-to-hex nil)
  counsel-colors--name-to-hex("alternateSelectedControlColor")
  #f(compiled-function (cell) #<bytecode 0x43c34a31>)(("alternateSelectedControlColor"))
  mapcar(#f(compiled-function (cell) #<bytecode 0x43c34a31>) (("LightGreen" "light green") ("DarkRed" "dark red") ("DarkMagenta" "dark magenta") ("DarkCyan" "dark cyan") ("DarkBlue" "dark blue") ("DarkGray" "dark grey" "DarkGrey" "dark gray") ("grey100" "gray100") ("grey99" "gray99") ("grey98" "gray98") ("grey97" "gray97") ("grey96" "gray96") ("grey95" "gray95") ("grey94" "gray94") ("grey93" "gray93") ("grey92" "gray92") ("grey91" "gray91") ("grey90" "gray90") ("grey89" "gray89") ("grey88" "gray88") ("grey87" "gray87") ("grey86" "gray86") ("grey85" "gray85") ("grey84" "gray84") ("grey83" "gray83") ("grey82" "gray82") ("grey81" "gray81") ("grey80" "gray80") ("grey79" "gray79") ("grey78" "gray78") ("grey77" "gray77") ("grey76" "gray76") ("grey75" "gray75") ("grey74" "gray74") ("grey73" "gray73") ("grey72" "gray72") ("grey71" "gray71") ("grey70" "gray70") ("grey69" "gray69") ("grey68" "gray68") ("grey67" "gray67") ("grey66" "gray66") ("grey65" "gray65") ("grey64" "gray64") ("grey63" "gray63") ("grey62" "gray62") ("grey61" "gray61") ("grey60" "gray60") ("grey59" "gray59") ("grey58" "gray58") ("grey57" "gray57") ("grey56" "gray56") ("grey55" "gray55") ("grey54" "gray54") ("grey53" "gray53") ("grey52" "gray52") ("grey51" "gray51") ("grey50" "gray50") ("grey49" "gray49") ("grey48" "gray48") ("grey47" "gray47") ("grey46" "gray46") ("grey45" "gray45") ("grey44" "gray44") ("grey43" "gray43") ("grey42" "gray42") ("grey41" "gray41") ("grey40" "gray40") ("grey39" "gray39") ("grey38" "gray38") ("grey37" "gray37") ("grey36" "gray36") ("grey35" "gray35") ("grey34" "gray34") ("grey33" "gray33") ("grey32" "gray32") ("grey31" "gray31") ("grey30" "gray30") ("grey29" "gray29") ("grey28" "gray28") ("grey27" "gray27") ("grey26" "gray26") ("grey25" "gray25") ("grey24" "gray24") ("grey23" "gray23") ("grey22" "gray22") ("grey21" "gray21") ("grey20" "gray20") ("grey19" "gray19") ("grey18" "gray18") ("grey17" "gray17") ("grey16" "gray16") ("grey15" "gray15") ("grey14" "gray14") ("grey13" "gray13") ("grey12" "gray12") ("grey11" "gray11") ("grey10" "gray10") ("grey9" "gray9") ("grey8" "gray8") ("grey7" "gray7") ("grey6" "gray6") ("grey5" "gray5") ("grey4" "gray4") ("grey3" "gray3") ("grey2" "gray2") ("grey1" "gray1") ("grey0" "gray0") ("thistle4") ("thistle3") ("thistle2") ("thistle1") ("MediumPurple4") ("MediumPurple3") ("MediumPurple2") ("MediumPurple1") ("purple4") ("purple3") ("purple2") ("purple1") ("DarkOrchid4") ("DarkOrchid3") ("DarkOrchid2") ("DarkOrchid1") ("MediumOrchid4") ("MediumOrchid3") ("MediumOrchid2") ("MediumOrchid1") ("plum4") ("plum3") ("plum2") ("plum1") ("orchid4") ("orchid3") ("orchid2") ("orchid1") ("magenta4") ("magenta3") ("magenta2") ("magenta1") ("VioletRed4") ("VioletRed3") ("VioletRed2") ("VioletRed1") ("maroon4") ("maroon3") ("maroon2") ("maroon1") ("PaleVioletRed4") ("PaleVioletRed3") ("PaleVioletRed2") ("PaleVioletRed1") ("LightPink4") ("LightPink3") ("LightPink2") ("LightPink1") ("pink4") ("pink3") ("pink2") ("pink1") ("HotPink4") ("HotPink3") ("HotPink2") ("HotPink1") ("DeepPink4") ("DeepPink3") ("DeepPink2") ("DeepPink1") ("red4") ("red3") ("red2") ("red1") ("OrangeRed4") ("OrangeRed3") ("OrangeRed2") ("OrangeRed1") ("tomato4") ("tomato3") ("tomato2") ("tomato1") ("coral4") ("coral3") ("coral2") ("coral1") ("DarkOrange4") ("DarkOrange3") ("DarkOrange2") ("DarkOrange1") ("orange4") ("orange3") ("orange2") ("orange1") ("LightSalmon4") ("LightSalmon3") ("LightSalmon2") ("LightSalmon1") ("salmon4") ("salmon3") ("salmon2") ("salmon1") ("brown4") ("brown3") ("brown2") ("brown1") ("firebrick4") ("firebrick3") ("firebrick2") ("firebrick1") ("chocolate4") ("chocolate3") ("chocolate2") ("chocolate1") ("tan4") ("tan3") ("tan2") ("tan1") ("wheat4") ("wheat3") ("wheat2") ("wheat1") ("burlywood4") ("burlywood3") ("burlywood2") ("burlywood1") ("sienna4") ("sienna3") ("sienna2") ("sienna1") ("IndianRed4") ("IndianRed3") ("IndianRed2") ("IndianRed1") ("RosyBrown4") ("RosyBrown3") ("RosyBrown2") ("RosyBrown1") ("DarkGoldenrod4") ("DarkGoldenrod3") ("DarkGoldenrod2") ("DarkGoldenrod1") ("goldenrod4") ("goldenrod3") ("goldenrod2") ("goldenrod1") ("gold4") ("gold3") ("gold2") ("gold1") ("yellow4") ("yellow3") ("yellow2") ("yellow1") ("LightYellow4") ("LightYellow3") ("LightYellow2") ("LightYellow1") ("LightGoldenrod4") ("LightGoldenrod3") ("LightGoldenrod2") ("LightGoldenrod1") ("khaki4") ("khaki3") ("khaki2") ("khaki1") ("DarkOliveGreen4") ("DarkOliveGreen3") ("DarkOliveGreen2") ("DarkOliveGreen1") ("OliveDrab4") ("OliveDrab3") ("OliveDrab2") ("OliveDrab1") ("chartreuse4") ("chartreuse3") ("chartreuse2") ("chartreuse1") ("green4") ("green3") ("green2") ("green1") ("SpringGreen4") ("SpringGreen3") ("SpringGreen2") ("SpringGreen1") ("PaleGreen4") ("PaleGreen3") ("PaleGreen2") ("PaleGreen1") ("SeaGreen4") ("SeaGreen3") ("SeaGreen2") ("SeaGreen1") ("DarkSeaGreen4") ("DarkSeaGreen3") ("DarkSeaGreen2") ("DarkSeaGreen1") ("aquamarine4") ("aquamarine3") ("aquamarine2") ("aquamarine1") ("DarkSlateGray4") ("DarkSlateGray3") ("DarkSlateGray2") ("DarkSlateGray1") ("cyan4") ("cyan3") ("cyan2") ("cyan1") ("turquoise4") ("turquoise3") ("turquoise2") ("turquoise1") ("CadetBlue4") ("CadetBlue3") ("CadetBlue2") ("CadetBlue1") ("PaleTurquoise4") ("PaleTurquoise3") ("PaleTurquoise2") ("PaleTurquoise1") ("LightCyan4") ("LightCyan3") ("LightCyan2") ("LightCyan1") ("LightBlue4") ("LightBlue3") ("LightBlue2") ("LightBlue1") ("LightSteelBlue4") ("LightSteelBlue3") ("LightSteelBlue2") ("LightSteelBlue1") ("SlateGray4") ("SlateGray3") ("SlateGray2") ("SlateGray1") ("LightSkyBlue4") ("LightSkyBlue3") ("LightSkyBlue2") ("LightSkyBlue1") ("SkyBlue4") ("SkyBlue3") ("SkyBlue2") ("SkyBlue1") ("DeepSkyBlue4") ("DeepSkyBlue3") ("DeepSkyBlue2") ("DeepSkyBlue1") ("SteelBlue4") ("SteelBlue3") ("SteelBlue2") ("SteelBlue1") ("DodgerBlue4") ("DodgerBlue3") ("DodgerBlue2") ("DodgerBlue1") ("blue4") ("blue3") ("blue2") ("blue1") ("RoyalBlue4") ("RoyalBlue3") ("RoyalBlue2") ("RoyalBlue1") ("SlateBlue4") ("SlateBlue3") ("SlateBlue2") ("SlateBlue1") ("azure4") ("azure3") ("azure2") ("azure1") ("MistyRose4") ("MistyRose3") ("MistyRose2") ("MistyRose1") ("LavenderBlush4") ("LavenderBlush3") ("LavenderBlush2") ("LavenderBlush1") ("honeydew4") ("honeydew3") ("honeydew2") ("honeydew1") ("ivory4") ("ivory3") ("ivory2") ("ivory1") ("cornsilk4") ("cornsilk3") ("cornsilk2") ("cornsilk1") ("LemonChiffon4") ("LemonChiffon3") ("LemonChiffon2") ("LemonChiffon1") ("NavajoWhite4") ("NavajoWhite3") ("NavajoWhite2") ("NavajoWhite1") ("PeachPuff4") ("PeachPuff3") ("PeachPuff2") ("PeachPuff1") ("bisque4") ("bisque3") ("bisque2") ("bisque1") ("AntiqueWhite4") ("AntiqueWhite3") ("AntiqueWhite2") ("AntiqueWhite1") ("seashell4") ("seashell3") ("seashell2") ("seashell1") ("snow4") ("snow3") ("snow2") ("snow1") ("thistle") ("MediumPurple" "medium purple") ("purple") ("BlueViolet" "blue violet") ("DarkViolet" "dark violet") ("DarkOrchid" "dark orchid") ("MediumOrchid" "medium orchid") ("orchid") ("plum") ("violet") ("magenta") ("VioletRed" "violet red") ("MediumVioletRed" "medium violet red") ("maroon") ("PaleVioletRed" "pale violet red") ("LightPink" "light pink") ("pink") ("DeepPink" "deep pink") ("HotPink" "hot pink") ("red") ("OrangeRed" "orange red") ("tomato") ("LightCoral" "light coral") ("coral") ("DarkOrange" "dark orange") ("orange") ("LightSalmon" "light salmon") ("salmon") ("DarkSalmon" "dark salmon") ("brown") ("firebrick") ("chocolate") ("tan") ("SandyBrown" "sandy brown") ("wheat") ("beige") ("burlywood") ("peru") ("sienna") ("SaddleBrown" "saddle brown") ("IndianRed" "indian red") ("RosyBrown" "rosy brown") ("DarkGoldenrod" "dark goldenrod") ("goldenrod") ("LightGoldenrod" "light goldenrod") ("gold") ("yellow") ("LightYellow" "light yellow") ("LightGoldenrodYellow" "light goldenrod yellow") ("PaleGoldenrod" "pale goldenrod") ("khaki") ("DarkKhaki" "dark khaki") ("OliveDrab" "olive drab") ("ForestGreen" "forest green") ("YellowGreen" "yellow green") ("LimeGreen" "lime green") ("GreenYellow" "green yellow") ("MediumSpringGreen" "medium spring green") ("chartreuse") ("green") ("LawnGreen" "lawn green") ("SpringGreen" "spring green") ("PaleGreen" "pale green") ("LightSeaGreen" "light sea green") ("MediumSeaGreen" "medium sea green") ("SeaGreen" "sea green") ("DarkSeaGreen" "dark sea green") ("DarkOliveGreen" "dark olive green") ("DarkGreen" "dark green") ("aquamarine") ("MediumAquamarine" "medium aquamarine") ("CadetBlue" "cadet blue") ("LightCyan" "light cyan") ("cyan") ("turquoise") ("MediumTurquoise" "medium turquoise") ("DarkTurquoise" "dark turquoise") ("PaleTurquoise" "pale turquoise") ("PowderBlue" "powder blue") ("LightBlue" "light blue") ("LightSteelBlue" "light steel blue") ("SteelBlue" "steel blue") ("LightSkyBlue" "light sky blue") ("SkyBlue" "sky blue") ("DeepSkyBlue" "deep sky blue") ("DodgerBlue" "dodger blue") ("blue") ("RoyalBlue" "royal blue") ("MediumBlue" "medium blue") ("LightSlateBlue" "light slate blue") ("MediumSlateBlue" "medium slate blue") ("SlateBlue" "slate blue") ("DarkSlateBlue" "dark slate blue") ("CornflowerBlue" "cornflower blue") ("NavyBlue" "navy" "navy blue") ("MidnightBlue" "midnight blue") ("LightGray" "light grey" "LightGrey" "light gray") ("grey" "gray") ("LightSlateGrey" "light slate gray" "LightSlateGray" "light slate grey") ("SlateGrey" "slate gray" "SlateGray" "slate grey") ("DimGrey" "dim gray" "DimGray" "dim grey") ("DarkSlateGrey" "dark slate gray" "DarkSlateGray" "dark slate grey") ("black") ("white") ("MistyRose" "misty rose") ("LavenderBlush" "lavender blush") ("lavender") ("AliceBlue" "alice blue") ("azure") ("MintCream" "mint cream") ("honeydew") ("seashell") ("LemonChiffon" "lemon chiffon") ("ivory") ("cornsilk") ("moccasin") ("NavajoWhite" "navajo white") ("PeachPuff" "peach puff") ("bisque") ("BlanchedAlmond" "blanched almond") ("PapayaWhip" "papaya whip") ("AntiqueWhite" "antique white") ("linen") ("OldLace" "old lace") ("FloralWhite" "floral white") ("gainsboro") ("WhiteSmoke" "white smoke") ("GhostWhite" "ghost white") ("snow") ("alternateSelectedControlColor") ("alternateSelectedControlTextColor" "controlColor" "controlBackgroundColor") ("controlDarkShadowColor" "controlShadowColor" "controlLightHighlightColor" "controlHighlightColor") ("controlTextColor" "disabledControlTextColor") ("gridColor") ("headerColor") ("headerTextColor") ("highlightColor") ("keyboardFocusIndicatorColor") ("knobColor") ("labelColor" "quaternaryLabelColor") ("scrollBarColor") ("secondaryLabelColor") ("secondarySelectedControlColor") ("selectedControlColor") ("selectedControlTextColor") ("selectedKnobColor" "selectedMenuItemColor") ("selectedMenuItemTextColor") ("selectedTextBackgroundColor") ("selectedTextColor") ("shadowColor") ("tertiaryLabelColor") ("textBackgroundColor") ("textColor") ("windowBackgroundColor") ("windowFrameColor") ("windowFrameTextColor" "Black") ("Blue") ("Brown") ("Cyan") ("Green") ("Magenta") ("Orange") ("Purple") ("Red") ("Yellow") ("White")))
  counsel-colors-emacs()
  funcall-interactively(counsel-colors-emacs)
  call-interactively(counsel-colors-emacs nil nil)
  command-execute(counsel-colors-emacs)
```

### PR DESCRIPTION
**Pull request by mookid**
fix #1853

### CODE REVIEW

File: counsel.el [hunk 1]
```diff
@@ -4459,12 +4463,15 @@ Return closure suitable for `ivy-format-function'."
...
 selected color."
   (interactive)
-  (let* ((colors (mapcar (lambda (cell)
-                           (let ((name (car cell)))
-                             (propertize name
-                                         'hex (counsel-colors--name-to-hex name)
-                                         'dups (cdr cell))))
-                         (list-colors-duplicates)))
+  (let* ((colors
+          (delete nil
```
Comment (basil-conto) [inline counsel.el]: Better to use `delq` here.

Comment (mookid) [reply counsel.el]: yes, in fact it is probably better not to allocate two lists and use mapcan instead.

Comment (basil-conto) [reply counsel.el]: It's not always clear when it's better to use `(delq nil (mapcar ...))` vs `mapcan`. The thing about `mapcan` is it was only added in Emacs 25, so you have to use the wrapper `cl-mapcan`, and you still need to allocate a cons cell in the loop for each element you want to keep:

```el
(cl-mapcan (lambda (cell)
             (let* ((name (car cell))
                    (dups (cdr cell))
                    (hex (counsel-colors--name-to-hex name)))
               (and hex (list (propertize name 'hex hex 'dups dups)))))
           (list-colors-duplicates))
```

Benchmarking the two approaches suggests `delq`+`mapcar` is a hair faster and smaller than `cl-mapcan`. I really don't mind either way, I'm just saying there are tradeoffs, so it basically comes down to personal preference.

Comment (mookid) [reply counsel.el]: Thanks for running the experimentation. 

Note that contrary to what I said, delete is destructive and does not make a new list (but remove does).

Comment (basil-conto) [reply counsel.el]: > Note that contrary to what I said, delete is destructive and does not make a new list (but remove does).

Yup. The difference between `delq` and `delete` is that the former only accepts lists and uses `eq` for equality, thus making it the standard function for use in `(delq nil (mapcar ...))`. Using `delete` is less common and reserved for cases where equivalence under `equal` is needed.

**Comment by abo-abo** [comment]
Thanks.

**Pull request merged**